dev/core#3692: use buildOptions for activityType, activityStatus where easy. Also a relationshipType

herbdool opened the pull request:
Fixes https://lab.civicrm.org/dev/core/-/issues/3692

This might be hard to test. I focused on the ones that were more straightforward and left the ones that might need more decisions.

herbdool committed:
Issue 3692: use buildOptions for activityType, activityStatus where easy. Also a relationshipType

demeritcowboy commented on CRM/Activity/BAO/Activity.php:
```diff
@@ -2384,7 +2384,7 @@ public static function getContactActivitySelector(&$params) {
 
     if (!empty($activities)) {
-      $activityStatus = CRM_Core_PseudoConstant::activityStatus();
+      $activityStatus = CRM_Activity_BAO_Activity::buildOptions('status_id', 'get');
```
These aren't quite the same. The original excludes disabled statuses, the replacement includes them. Ideally, we'd review each one carefully to see if the original made sense or whether it should include disabled, but then that will stall. (In this particular spot, the before gives an E_NOTICE and to me looks wrong because in the selector it shows as no status, so I like the replacement. But others will disagree.)

herbdool replied:
Ah yes. Looking deeper into the code, I think we can use 'search' instead of 'get' for status. The only change seems to be it excludes inactive options and inactive components' options. At least for activity status and type.

herbdool replied:
I've changed this to 'search' and in a few other places.

herbdool committed:
fix some of the status gets to searches and fix some more activity types

herbdool commented:
I made another pass. I changed some of the "get" to "search" for fetching statuses, where the original code was not including inactive items. And I fixed a few others. In all cases I tried to ensure the original filtering was preserved. Where fetching an array of activity types or statuses was only used to find the label I either left it to fetch all, or switched it to use `CRM_Core_Pseudoconstant::getLabel()`. 

There were a couple tricky spots where it either needed to only include case activities or exclude. For that I used `CRM_Core_OptionGroup::values()` which I saw was being used in a report already. Though it is quite low level I didn't see any warnings about using it or not. And it appears to be cached.

For reference the original `CRM_Core_PseudoConstant::activityType()` uses this to include/exclude:

```
$all, defaults to TRUE
$includeCaseActivities, defaults to FALSE
$reset, defaults to FALSE
$returnColumn, defaults to 'label'
$includeCampaignActivities, defaults to FALSE
$onlyComponentActivities, defaults to FALSE
```

In each case I went over the flags and ensured they would still match. The replacement `buildOptions` doesn't care about including case/campaign/component activities - they're included by default with 'get' and with 'search' are excluded only if they are disabled or if the component is disabled.

demeritcowboy commented on CRM/Report/Form/Activity.php:
```diff
@@ -564,7 +564,8 @@ public function where($recordType = NULL) {
                 if (empty($this->_params['include_case_activities_value'])) {
-                  $this->activityTypes = CRM_Core_PseudoConstant::activityType(TRUE, FALSE, FALSE, 'label', TRUE);
+                  $componentId = CRM_Core_Component::getComponentID('CiviCase');
+                  $this->activityTypes = CRM_Core_OptionGroup::values('activity_type', FALSE, FALSE, FALSE, " AND v.component_id !={$componentId} OR v.component_id IS NULL");
```
All of the test fails seem to be here. Because $componentId is null if civicase is enabled.

[CRM_Report_Form_ActivityTest.testLongCustomFieldNames](https://test.civicrm.org/job/CiviCRM-Core-PR/49961/testReport/junit/(root)/CRM_Report_Form_ActivityTest/testLongCustomFieldNames/)
    [CRM_Report_Form_ActivityTest.testTargetAddressFields](https://test.civicrm.org/job/CiviCRM-Core-PR/49961/testReport/junit/(root)/CRM_Report_Form_ActivityTest/testTargetAddressFields/)
    [api_v3_ReportTemplateTest.testActivityDetails](https://test.civicrm.org/job/CiviCRM-Core-PR/49961/testReport/junit/(root)/api_v3_ReportTemplateTest/testActivityDetails/)
    [api_v3_ReportTemplateTest.testActivityDetailsNullFilters](https://test.civicrm.org/job/CiviCRM-Core-PR/49961/testReport/junit/(root)/api_v3_ReportTemplateTest/testActivityDetailsNullFilters/)
    [api_v3_ReportTemplateTest.testActivityDetailsContactFilter](https://test.civicrm.org/job/CiviCRM-Core-PR/49961/testReport/junit/(root)/api_v3_ReportTemplateTest/testActivityDetailsContactFilter/)
    [api_v3_ReportTemplateTest.testReportTemplateSelectWhere with data set #2](https://test.civicrm.org/job/CiviCRM-Core-PR/49961/testReport/junit/(root)/api_v3_ReportTemplateTest/testReportTemplateSelectWhere_with_data_set__2/)
    [api_v3_ReportTemplateTest.testReportTemplateGetRowsAllReports with data set #2](https://test.civicrm.org/job/CiviCRM-Core-PR/49961/testReport/junit/(root)/api_v3_ReportTemplateTest/testReportTemplateGetRowsAllReports_with_data_set__2/)
    [api_v3_ReportTemplateTest.testReportTemplateGetRowsAllReportsACL with data set #2](https://test.civicrm.org/job/CiviCRM-Core-PR/49961/testReport/junit/(root)/api_v3_ReportTemplateTest/testReportTemplateGetRowsAllReportsACL_with_data_set__2/)
    [api_v3_ReportTemplateTest.testReportTemplateGetStatisticsAllReports with data set #2](https://test.civicrm.org/job/CiviCRM-Core-PR/49961/testReport/junit/(root)/api_v3_ReportTemplateTest/testReportTemplateGetStatisticsAllReports_with_data_set__2/)

herbdool replied:
Thanks. I'll see if my update fixes that.

herbdool committed:
don't check components if CiviCase not enabled

demeritcowboy commented on CRM/Activity/BAO/Activity.php:
```diff
@@ -2511,7 +2509,7 @@ public static function getContactActivitySelector(&$params) {
 
         $activity['activity_date_time'] = CRM_Utils_Date::customFormat($values['activity_date_time']);
-        $activity['status_id'] = $activityStatus[$values['status_id']];
+        $activity['status_id'] = CRM_Core_Pseudoconstant::getLabel('CRM_Activity_BAO_Activity', 'status_id', $values['status_id']);
```
These aren't the same - it now includes disabled. I personally like that but some people won't.

demeritcowboy commented on CRM/Activity/Form/Activity.php:
```diff
@@ -143,7 +143,7 @@ public function getDefaultEntity() {
     // Remove print document activity type
     $unwanted = CRM_Core_OptionGroup::values('activity_type', FALSE, FALSE, FALSE, "AND v.name = 'Print PDF Letter'");
-    $activityTypes = array_diff_key(CRM_Core_PseudoConstant::ActivityType(FALSE), $unwanted);
+    $activityTypes = array_diff_key(CRM_Activity_BAO_Activity::buildOptions('activity_type_id', 'search'), $unwanted);
```
These aren't the same. Before it didn't include case activity types and now it does and also all the ones with filter=1.

demeritcowboy commented on tests/phpunit/Civi/CCase/SequenceListenerTest.php:
```diff
@@ -25,7 +25,7 @@ public function setUp(): void {
 
   public function testSequence() {
-    $actStatuses = array_flip(\CRM_Core_PseudoConstant::activityStatus('name'));
+    $actStatuses = array_flip(\CRM_Activity_BAO_Activity::buildOptions('status_id', 'validate'));
```
This isn't the same, but for it's use here in the test is ok.

eileenmcnaughton commented:
Unfortunately this has gone stale & needs a rebase - I pulled out a handful of small changes into separate commits so this will lose a bit of bulk when it is rebased

herbdool commented:
I've rebased this. Though would need to check a few more comments before it's ready.

civicrm-builder commented:
Can one of the admins verify this patch?

eileenmcnaughton commented on CRM/Campaign/Form/Task/Reserve.php:
```diff
@@ -70,7 +70,7 @@ public function preProcess() {
     CRM_Campaign_BAO_Survey::retrieve($params, $this->_surveyDetails);
 
     //get the survey activities.
-    $activityStatus = CRM_Core_PseudoConstant::activityStatus('name');
```
@demeritcowboy these swaps are safe aren't they? This PR is stale & I think turned out to be way trickier than initial appearance but I'm thinking maybe I can salvage some of the bits that are straight foward

demeritcowboy replied:
Some of them might be.

eileenmcnaughton commented:
I'm going to close this out because it's stale - it probably stalled on some of the challenges picked up in the review. I did save the last couple of changes in https://github.com/civicrm/civicrm-core/pull/28384 (just because I randomly started from the bottom)
